### PR TITLE
Add Api data to db

### PR DESCRIPTION
**Pull request by CressChe**
https://trello.com/c/f3TDgf8A/31-add-api-data-to-db

### CODE REVIEW

File: WhaleSpotting.UnitTests/Services/SightingsServiceTests.cs [hunk 1]
```diff
@@ -41,6 +43,43 @@ public async Task GetSightings_Called_ReturnsSightings()
...
+            // Assert
+            result.Should().BeOfType<List<SightingResponseModel>>();
+            result.Should().HaveCount(2);
+
+            var whaleSightingsDbModels = Context.Sightings.Where(s => s.Description == whaleSighting.Description).ToList();
+            whaleSightingsDbModels.Should().HaveCount(2);
+            whaleSightingsDbModels.Should().BeOfType<List<SightingDbModel>>();
+        }
+    
+
```
Comment (oskarwilliams) [inline WhaleSpotting.UnitTests/Services/SightingsServiceTests.cs]: Double new line

File: WhaleSpotting/Controllers/ApiDataController.cs [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    public class ApiDataController : ControllerBase
+    {
+        private readonly ISightingsService _sightings;
+
+        public ApiDataController(ISightingsService sightings)
+        {
+            _sightings = sightings;
+        }
+
+        [HttpPost("")]
```
Comment (oskarwilliams) [inline WhaleSpotting/Controllers/ApiDataController.cs]:
```suggestion
        [HttpPost]
```

File: WhaleSpotting/Controllers/ApiDataController.cs [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+            var page = 1;
+            var hasResults = true;
+            var sightingsToAdd = new List<SightingDbModel>();
+            var client = new RestClient("https://hotline.whalemuseum.org/");
+
+            while (hasResults)
+            {
+                var request = new RestRequest($"api.json?limit=1000&page={page}", DataFormat.Json);
+                var apiSightings = await client.GetAsync<List<SightingApiModel>>(request);
+                if (apiSightings.Any())
```
Comment (oskarwilliams) [inline WhaleSpotting/Controllers/ApiDataController.cs:36]: New line before this please

File: WhaleSpotting/Controllers/ApiDataController.cs [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Mvc;
+using System.Text;
+using System.Net;
+using RestSharp;
+using WhaleSpotting.Models.ApiModels;
+using WhaleSpotting.Models.DbModels;
+using WhaleSpotting.Services;
+using Microsoft.AspNetCore.Authorization;
```
Comment (oskarwilliams) [inline WhaleSpotting/Controllers/ApiDataController.cs]: A load of these aren't needed

File: WhaleSpotting/Controllers/ApiDataController.cs [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    {
+        private readonly ISightingsService _sightings;
+
+        public ApiDataController(ISightingsService sightings)
+        {
+            _sightings = sightings;
+        }
+
+        [HttpPost("")]
+        public async Task<OkResult> GetApiData()
```
Comment (oskarwilliams) [inline WhaleSpotting/Controllers/ApiDataController.cs]: I'd prefer this be with the IActionResult
```suggestion
        public async Task<IActionResult> GetApiData()
```

File: WhaleSpotting/Migrations/20210804101051_editedSightingsDbModel.cs [hunk 6]
```diff
@@ -0,0 +1,23 @@
+﻿using Microsoft.EntityFrameworkCore.Migrations;
+
+namespace WhaleSpotting.Migrations
+{
+    public partial class editedSightingsDbModel : Migration
```
Comment (oskarwilliams) [inline WhaleSpotting/Migrations/20210804101051_editedSightingsDbModel.cs]: Could we regenerate this migration with a capital letter start. We use pascal case for migrations :)

File: WhaleSpotting/Models/ApiModels/SightingApiModel.cs [hunk 7]
```diff
@@ -0,0 +1,94 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
```
Comment (oskarwilliams) [inline WhaleSpotting/Models/ApiModels/SightingApiModel.cs]: This isn't needed

File: WhaleSpotting/Models/DbModels/SightingDbModel.cs [hunk 8]
```diff
@@ -1,15 +1,19 @@
...
 using WhaleSpotting.Models.Enums;
 
 namespace WhaleSpotting.Models.DbModels
 {
     public class SightingDbModel
     {
         public int Id { get; set; }
-        public Species Species { get; set; }
-        public int Quantity { get; set; }
+
```
Comment (oskarwilliams) [inline WhaleSpotting/Models/DbModels/SightingDbModel.cs]: Random new line :)

File: WhaleSpotting/Models/DbModels/SightingDbModel.cs [hunk 9]
```diff
@@ -21,7 +25,9 @@ public class SightingDbModel
         public string OrcaPod { get; set; }
         public bool Confirmed { get; set; }
 
-        // TODO - Add UserDbModel
-        //public UserDbModel User {get; set; }
-    }
+        public SightingDbModel ()
+        {
+
+        }
```
Comment (oskarwilliams) [inline WhaleSpotting/Models/DbModels/SightingDbModel.cs]: Can we put the TODO back and remove this constructor

File: WhaleSpotting/Models/DbModels/SightingDbModel.cs [hunk 10]
```diff
@@ -1,15 +1,19 @@
 ﻿using System;
 using System.Collections.Generic;
 using System.ComponentModel.DataAnnotations.Schema;
+using System.Text.RegularExpressions;
```
Comment (oskarwilliams) [inline WhaleSpotting/Models/DbModels/SightingDbModel.cs]: A lot of these aren't needed

File: WhaleSpotting/Program.cs [hunk 11]
```diff
@@ -2,21 +2,22 @@
...
+using WhaleSpotting.Controllers;
 using WhaleSpotting.Models.DbModels;
 
+
 namespace WhaleSpotting
 {
     public class Program
     {
-        public static void Main(string[] args)
+        public static async System.Threading.Tasks.Task Main(string[] args)
```
Comment (oskarwilliams) [inline WhaleSpotting/Program.cs]: Can we revert this files changes pls

File: WhaleSpotting/Services/SightingsService.cs [hunk 12]
```diff
@@ -36,6 +37,27 @@ public async Task<List<SightingResponseModel>> GetSightings()
...
+            var sightingsNotInDb =
+                sightingsToAdd.Where(s => !sightingsInDb.Contains(s.ApiId));
+            sightingsNotInDb
+                .ToList();
+      
+            _context.Sightings.AddRange(sightingsNotInDb.ToArray());
+            _context.SaveChanges();
+
+            return sightingsNotInDb.Select(s => new SightingResponseModel(s)).ToList();
+        }
```
Comment (oskarwilliams) [inline WhaleSpotting/Services/SightingsService.cs]:
```suggestion
        public List<SightingResponseModel> CreateSightings(List<SightingDbModel> sightingsToAdd)
        {
            var newSightingIds = sightingsToAdd.Select(s => s.ApiId).Distinct();
            var sightingsInDbIds = _context.Sightings
                .Where(s => newSightingIds.Contains(s.ApiId))
                .Select(s => s.ApiId);

            var sightingsNotInDb = sightingsToAdd
                .Where(s => !sightingsInDbIds.Contains(s.ApiId))
                .ToList();

            _context.Sightings.AddRange(sightingsNotInDb);
            _context.SaveChanges();

            return sightingsNotInDb.Select(s => new SightingResponseModel(s)).ToList();
        }
```

File: WhaleSpotting.UnitTests/Services/SightingsServiceTests.cs [hunk 13]
```diff
@@ -41,6 +43,43 @@ public async Task GetSightings_Called_ReturnsSightings()
...
+            {
+                
+                Species = "orca",
+                Quantity = "50",
+                Location = "Southend",
+                Latitude = 48.6213,
+                Longitude = -123.2828,
+                Description = "Sighted near lighthouse",
+                SightedAt = System.DateTime.Now,
+                CreatedAt = System.DateTime.Now,
```
Comment (oskarwilliams) [inline WhaleSpotting.UnitTests/Services/SightingsServiceTests.cs]:
```suggestion
                SightedAt = DateTime.Now,
                CreatedAt = DateTime.Now,
```

File: WhaleSpotting.UnitTests/WhaleSpotting.UnitTests.csproj [hunk 14]
```diff
@@ -16,6 +16,7 @@
           <PrivateAssets>all</PrivateAssets>
           <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
         </PackageReference>
+        <PackageReference Include="Newtonsoft.Json" Version="13.0.1" />
```
Comment (oskarwilliams) [inline WhaleSpotting.UnitTests/WhaleSpotting.UnitTests.csproj]: Is this needed? Can we remove this?

**Review by oskarwilliams — CHANGES_REQUESTED** [review]

File: WhaleSpotting.UnitTests/Services/SightingsServiceTests.cs (hunk 1 above)
Comment (CressChe) [reply WhaleSpotting.UnitTests/Services/SightingsServiceTests.cs]: removed line

File: WhaleSpotting/Controllers/ApiDataController.cs (hunk 2 above)
Comment (CressChe) [reply WhaleSpotting/Controllers/ApiDataController.cs]: :)

File: WhaleSpotting/Controllers/ApiDataController.cs (hunk 3 above)
Comment (CressChe) [reply WhaleSpotting/Controllers/ApiDataController.cs:36]: added line

File: WhaleSpotting/Controllers/ApiDataController.cs (hunk 4 above)
Comment (CressChe) [reply WhaleSpotting/Controllers/ApiDataController.cs]: removed

File: WhaleSpotting/Controllers/ApiDataController.cs (hunk 5 above)
Comment (CressChe) [reply WhaleSpotting/Controllers/ApiDataController.cs]: :)

File: WhaleSpotting/Models/DbModels/SightingDbModel.cs (hunk 10 above)
Comment (CressChe) [reply WhaleSpotting/Models/DbModels/SightingDbModel.cs]: removed

File: WhaleSpotting/Program.cs (hunk 11 above)
Comment (CressChe) [reply WhaleSpotting/Program.cs]: done

File: WhaleSpotting/Services/SightingsService.cs (hunk 12 above)
Comment (CressChe) [reply WhaleSpotting/Services/SightingsService.cs]: :)

File: WhaleSpotting.UnitTests/Services/SightingsServiceTests.cs (hunk 13 above)
Comment (CressChe) [reply WhaleSpotting.UnitTests/Services/SightingsServiceTests.cs]: done

File: WhaleSpotting/Models/ApiModels/SightingApiModel.cs (hunk 7 above)
Comment (CressChe) [reply WhaleSpotting/Models/ApiModels/SightingApiModel.cs]: removed System.Linq

File: WhaleSpotting/Models/DbModels/SightingDbModel.cs (hunk 8 above)
Comment (CressChe) [reply WhaleSpotting/Models/DbModels/SightingDbModel.cs]: removed line

File: WhaleSpotting/Models/DbModels/SightingDbModel.cs (hunk 9 above)
Comment (CressChe) [reply WhaleSpotting/Models/DbModels/SightingDbModel.cs]: :)

File: WhaleSpotting/Migrations/20210804101051_editedSightingsDbModel.cs (hunk 6 above)
Comment (CressChe) [reply WhaleSpotting/Migrations/20210804101051_editedSightingsDbModel.cs]: done :)

File: WhaleSpotting.UnitTests/Services/SightingsServiceTests.cs [hunk 15]
```diff
@@ -41,6 +43,42 @@ public async Task GetSightings_Called_ReturnsSightings()
...
 
+        [Fact]
+        public void CreateSightings_CalledWithSightingsApiModel_ReturnListSightingResponseModelAndAddsToDb()
+        {
+            // Arrange
+            var sightingToAdd = new List<SightingDbModel>();
+
+            var whaleSighting = new SightingApiModel
+            {
+
```
Comment (oskarwilliams) [inline WhaleSpotting.UnitTests/Services/SightingsServiceTests.cs]: Extra new line

File: WhaleSpotting.UnitTests/WhaleSpotting.UnitTests.csproj (hunk 14 above)
Comment (oskarwilliams) [reply WhaleSpotting.UnitTests/WhaleSpotting.UnitTests.csproj]: Bump

File: WhaleSpotting/Program.cs [hunk 16]
```diff
@@ -2,8 +2,10 @@
 using Microsoft.EntityFrameworkCore;
 using Microsoft.Extensions.DependencyInjection;
 using Microsoft.Extensions.Hosting;
+using WhaleSpotting.Controllers;
 using WhaleSpotting.Models.DbModels;
 
+
```
Comment (oskarwilliams) [inline WhaleSpotting/Program.cs]: Extra new line and import still

File: WhaleSpotting/Services/SightingsService.cs [hunk 17]
```diff
@@ -36,6 +36,24 @@ public async Task<List<SightingResponseModel>> GetSightings()
             return sightings;
         }
 
+        public List<SightingResponseModel> CreateSightings(List<SightingDbModel> sightingsToAdd)
+        {
+            var newSightingIds = sightingsToAdd.Select(s => s.ApiId).Distinct();
+           
+            var sightingsInDbIds = _context.Sightings
+                                   .Where(s => newSightingIds.Contains(s.ApiId))
```
Comment (oskarwilliams) [inline WhaleSpotting/Services/SightingsService.cs]: These are a bit too far indented I think, should just be 1 tab

**Review by oskarwilliams — CHANGES_REQUESTED** [review]

File: WhaleSpotting.UnitTests/WhaleSpotting.UnitTests.csproj (hunk 14 above)
Comment (CressChe) [reply WhaleSpotting.UnitTests/WhaleSpotting.UnitTests.csproj]: :)

File: WhaleSpotting.UnitTests/Services/SightingsServiceTests.cs (hunk 15 above)
Comment (CressChe) [reply WhaleSpotting.UnitTests/Services/SightingsServiceTests.cs]: :)

File: WhaleSpotting/Program.cs (hunk 16 above)
Comment (CressChe) [reply WhaleSpotting/Program.cs]: done

File: WhaleSpotting/Services/SightingsService.cs (hunk 17 above)
Comment (CressChe) [reply WhaleSpotting/Services/SightingsService.cs]: :)

**Review by oskarwilliams — APPROVED** [review]

**Review by oskarwilliams — APPROVED** [review]

**Pull request merged**